plugin/iplookup/cache.go

```diff
@@ -0,0 +1,119 @@
+package iplookup
+
+import (
+	"net/http"
+	"time"
+)
+
+// Server the cache
+func (ipl *IPLookup) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	lookup := r.URL.String()
+	if len(lookup) > 0 && lookup[0] == '/' {
+		lookup = lookup[1:]
+	}
+
+	ipl.Lock()
+	entry, _ := ipl.cacheLookup[lookup]
+	ipl.Unlock()
+
+	if entry == nil {
+		http.NotFound(w, r)
+		return
+	}
+
+	// Write the entry
+	w.Write([]byte(entry.value))
+}
+
+func (ipl *IPLookup) addCache(name, value string) {
+
```

Comment:
This seems like an attempt not to use defer but the code would be simpler like:
```
func (ipl *IPLookup) addCache(name, value string) {
 ipl.Lock()
 defer ipl.Unlock()

 // See if we know about this entry
 if entry, found := ipl.cacheLookup[name]; found {
  // Update the expires time
  entry.expires = time.Now().Add(ipl.maxDuration)

  // We only need to modify if we're not the head already
  if entry.next == nil {
   break
  }

  // We are the tail, we know we're not the head
  if entry.prev == nil {
   ipl.tail = entry.next
   entry.next.prev = nil
  } else {
   // We know we're not the tail and not the head
   entry.prev.next = entry.next
   entry.next.prev = entry.prev
  }

  ipl.head.next = entry
  entry.prev = ipl.head
  entry.next = nil
  ipl.head = entry
  return
 }

 // Doesn't exist, add it to the cache
 entry = &cacheEntry{
  name:    name,
  value:   value,
  expires: time.Now().Add(ipl.maxDuration),
  next: nil,
  prev: ipl.head,
 }

 if ipl.head != nil {
  ipl.head.next = entry
 } else {
  ipl.tail = entry // Head is nil, so is tail, set it
 }

 ipl.head = entry
 ipl.cacheLookup[name] = entry

 ipl.size++
}
```

Reply:
It definitely was so that I did not have to use defer. There's several articles that talk about using defer vs not using it. I think the use of defer makes lots of sense if you have a lot of control logic leading to multiple possible return points in a program and you don't want to sprinkle Unlock or Close throughout the function but when there's a simple case of of Lock when you start, Unlock when you leave, it will actually perform (albeit slightly) better if you inline the calls. This function will be called a lot and should remain performant as possible.